### PR TITLE
RAIL-51 Group fares in a new way

### PR DESCRIPTION
**Pull request by Dyktus**

### CODE REVIEW

File: src/journey/JourneyPlanner.ts [hunk 1]
```diff
@@ -56,7 +57,14 @@ export class JourneyPlanner {
...
         for (const journey of paginatedJourneys) {
-            journeyFareMap[journey.id] = [factory.getSingle(journey, fares.outwardSingles)];
+            if (!journeyFareMap[journey.id]) {
+                journeyFareMap[journey.id] = {
+                    'outward': [],
+                    'inward': [],
+                    'return': [],
+                };
+            }
+            journeyFareMap[journey.id]['outward'].push(factory.getSingle(journey, fares.outwardSingles));
```
Comment (Dyktus) [inline src/journey/JourneyPlanner.ts]: Question: Do we also need grouping here if we use only outwardSingles?

File: src/journey/JourneyPlanner.ts [hunk 2]
```diff
@@ -90,25 +106,63 @@ export class JourneyPlanner {
...
+                const returns = Object.keys(returnGroups)
+                    .map(group => factory.getReturn(outJourney, inJourney, returnGroups[group]))
+                    .filter(fare => fare.isDefined);
+                journeyFareMap[outJourney.id][inJourney.id] = {
+                    outward: outwardSingles,
+                    inward: inwardSingles,
+                    return: returns,
+                };
+
+                // Find max return price
```
Comment (Dyktus) [inline src/journey/JourneyPlanner.ts]: Question: Do we still need to find max return price or we should return everything what is in lines above (130-134) and remove that part of code below?

File: src/journey/JourneyPlanner.ts [hunk 3]
```diff
@@ -123,6 +177,17 @@ export class JourneyPlanner {
         return this.getJourneyPlanResponse(outwardPaginatedJourneys, inwardPaginatedJourneys, journeyFareMap, factory.cache, paginationResults);
     }
 
+    private groupFares(fares: SortedFareList): object {
+        const group = {};
```
Comment (Dyktus) [inline src/journey/JourneyPlanner.ts]: If I can use `reduce` here can you show me how? I think that I don't feel this method :)

File: src/journey/JourneyPlanner.ts [hunk 4]
```diff
@@ -56,7 +57,14 @@ export class JourneyPlanner {
         const journeyFareMap = {};
 
         for (const journey of paginatedJourneys) {
-            journeyFareMap[journey.id] = [factory.getSingle(journey, fares.outwardSingles)];
+            if (!journeyFareMap[journey.id]) {
```
Comment (linusnorton) [inline src/journey/JourneyPlanner.ts]: Why do you need this if statement?

File: src/journey/JourneyPlanner.ts (hunk 1 above)
Comment (linusnorton) [reply src/journey/JourneyPlanner.ts]: yes

File: src/journey/JourneyPlanner.ts (hunk 3 above)
Comment (orzeuek) [reply src/journey/JourneyPlanner.ts]: `(fares: SortedFareList): object => {

    return fares.reduce((carry: object, fare:Fare) => {
        if (!carry[this.getGroupKey(fare)]) {
            carry[this.getGroupKey(fare)] = [];
        }
        carry[this.getGroupKey(fare)].push(fare);
        
        return carry;
    }, {});
};`

File: src/journey/JourneyPlanner.ts [hunk 5]
```diff
@@ -90,25 +106,63 @@ export class JourneyPlanner {
...
+        // Grouping fares
+        const outwardSingleGroups = this.groupFares(fares.outwardSingles);
+        const inwardSingleGroups = this.groupFares(fares.inwardSingles);
+        const returnGroups = this.groupFares(fares.returns);
+
         for (const outJourney of outwardPaginatedJourneys) {
-            const outSingle = factory.getSingle(outJourney, fares.outwardSingles);
+            const outwardSingles = Object.keys(outwardSingleGroups)
+                .map(group => factory.getSingle(outJourney, outwardSingleGroups[group]))
+                .filter(fare => fare.isDefined);
```
Comment (linusnorton) [inline src/journey/JourneyPlanner.ts]: Good

File: src/journey/JourneyPlanner.ts (hunk 2 above)
Comment (linusnorton) [reply src/journey/JourneyPlanner.ts]: remove things below. I will need to do something but I'll sort it out later

File: src/journey/JourneyPlanner.ts (hunk 3 above)
Comment (linusnorton) [reply src/journey/JourneyPlanner.ts]: `fares.reduce(groupBy(f => this.getGroupKey(f)), {});`

File: src/service/api/JourneyPlanResponse.ts [hunk 6]
```diff
@@ -76,52 +76,37 @@ export class JourneyPlanResponse {
...
-                : this.journeyFareMapToJSON(fares[journeyId] as JourneyFareMap);
-
+            if (
+                !fares[journeyId].hasOwnProperty('outward') &&
+                !fares[journeyId].hasOwnProperty('inward') &&
+                !fares[journeyId].hasOwnProperty('return')
+            ) {
+                return this.journeyFareMapToJSON(fares[journeyId] as JourneyFareMap);
+            }
+            if (!result[journeyId]) {
```
Comment (linusnorton) [inline src/service/api/JourneyPlanResponse.ts]: This shouldn't be necessary

File: src/service/api/JourneyPlanResponse.ts [hunk 7]
```diff
@@ -76,52 +76,37 @@ export class JourneyPlanResponse {
...
+                result[journeyId] = {
+                    outward: [],
+                    inward: [],
+                    return: [],
+                };
+            }
+            for (const key of ['outward', 'inward', 'return']) {
+                if (fares[journeyId][key].hasOwnProperty('length')) {
```
Comment (linusnorton) [inline src/service/api/JourneyPlanResponse.ts]: confused by these two lines. Can you do `for (const key of Object.keys(result[journeyId]))` instead?

File: src/journey/JourneyPlanner.ts (hunk 3 above)
Comment (orzeuek) [reply src/journey/JourneyPlanner.ts]: only because you already implemented `groupBy` ! - that's cheating :D

Comment (Dyktus) [reply src/journey/JourneyPlanner.ts]: Is that some tournament "who will write it shorter"? :D

Comment (linusnorton) [reply src/journey/JourneyPlanner.ts]: The aim is not shorter! The aim is most descriptive/user intuitive. Yes I certainly did cheat by having that method already exist, but you guys should be aware that lib is already included and quite handy. 

It also has flatten, pushNested and setNested which are quite widely used.

File: src/journey/JourneyPlanner.ts [hunk 8]
```diff
@@ -55,21 +57,36 @@ export class JourneyPlanner {
         const factory = this.fareOptionFactory.create(request.passengerSet, availability);
         const journeyFareMap = {};
 
+        const groupOfFares = fares.outwardSingles.reduce(groupBy(this.getGroupKey), {});
         for (const journey of paginatedJourneys) {
-            journeyFareMap[journey.id] = [factory.getSingle(journey, fares.outwardSingles)];
+            if (!journeyFareMap[journey.id]) {
```
Comment (linusnorton) [inline src/journey/JourneyPlanner.ts]: what is this if statement here for?

File: src/journey/JourneyPlanner.ts [hunk 9]
```diff
@@ -55,21 +57,36 @@ export class JourneyPlanner {
...
+                journeyFareMap[journey.id] = {
+                    'outward': [],
```
Comment (linusnorton) [inline src/journey/JourneyPlanner.ts]: why initialise if you are just going to blat it later anyway?

**Pull request merged**